Replace MMI code fence with callbacks and arguments

vinistevam opened the pull request:
## **Description**
This PR aims to replace MMI code fences with callbacks and arguments to prepare (subsequent ticket) the support of the code in the Transaction Controller in the core repo as it is a static package and cannot have any build-specific code.

This effort is part of the unification of the Transaction Controller.

## **Manual testing steps**
**MM**
Regression test for transactions

## **Related issues**

Fixes https://github.com/MetaMask/MetaMask-planning/issues/997

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained:
  - [x] What problem this PR is solving.
  - [x] How this problem was solved.
  - [x] How reviewers can test my changes.
- [x] I’ve indicated what issue this PR is linked to: Fixes #???
- [x] I’ve included tests if applicable.
- [x] I’ve documented any added code.
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
- [x] I’ve properly set the pull request status:
  - [x] In case it's not yet "ready for review", I've set it to "draft".
  - [x] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

vinistevam committed:
replace mmi code fence

vinistevam committed:
add callback to custodyStatus

vinistevam committed:
add no-unused-vars

vinistevam commented:
> Any token transaction performed from test-dapp causes the error while tokens succesfully created and transferred.
> 
>  mmi.mov

It's better to perform a regression around transactions in the mm build. I spoke with MMI and they are having some issues due to an external library. Before merging it we can try again to test the MMI build.

sleepytanya commented:
Regression tests are passing except this [bug](https://github.com/MetaMask/metamask-extension/issues/21261).

matthewwalsh0 commented on app/scripts/controllers/transactions/index.js:
```diff
@@ -1471,24 +1471,17 @@ export default class TransactionController extends EventEmitter {
-      ///: BEGIN:ONLY_INCLUDE_IN(build-mmi)
-      txMeta.custodyStatus ? txMeta : undefined,
-      ///: END:ONLY_INCLUDE_IN
+      this._shouldDisablePublish(txMeta) ? txMeta : undefined,
```
We're using this single callback in multiple places as the condition for all the existing MMI logic, but that means the controller still has to know the MMI specific logic but use a boolean callback to trigger it, so in essence what we have is a `isMMI` function we're using in multiple places.

Ideally what we want is completely decoupled logic injected through multiple callbacks.

This one for example could be an optional `getAdditionalSignArguments` callback as then we can support any arguments here rather than just the metadata which is what MMI wants.

vinistevam replied:
Created the `getAdditionalSignArguments` supporting any argument.

matthewwalsh0 commented on app/scripts/controllers/transactions/index.js:
```diff
@@ -1471,24 +1471,17 @@ export default class TransactionController extends EventEmitter {
-      txMeta.custodyId = signedEthTx.custodian_transactionId;
-      txMeta.custodyStatus = signedEthTx.transactionStatus;
-
+    if (this._shouldDisablePublish(txMeta, signedEthTx)) {
```
This is the true should publish check, but as it has logic also maybe a better name would be an optional `beforePublish`?

It should update the metadata, and invoke the `transactionUpdateController` inside the callback, and then we can use the boolean return value of that callback to know whether to `return` or not.

matthewwalsh0 commented on app/scripts/controllers/transactions/index.js:
```diff
@@ -1471,24 +1471,17 @@ export default class TransactionController extends EventEmitter {
-      txMeta.custodyId = signedEthTx.custodian_transactionId;
-      txMeta.custodyStatus = signedEthTx.transactionStatus;
-
+    if (this._shouldDisablePublish(txMeta, signedEthTx)) {
```
Would this be cleaner as `afterSign` and we'd use the callback `boolean` return value to only return here if `false`.

This callback should only contain the fenced logic so settings the custody properties in the metadata and the call to the `transactionUpdateController`.

vinistevam replied:
Perfect, created a new optional callback with the fenced code presented in the `transaction/index.js`.

matthewwalsh0 commented on app/scripts/controllers/transactions/index.js:
```diff
@@ -172,6 +173,7 @@ export default class TransactionController extends EventEmitter {
     this.securityProviderRequest = opts.securityProviderRequest;
     this.messagingSystem = opts.messenger;
     this._hasCompletedOnboarding = opts.hasCompletedOnboarding;
+    this._shouldDisablePublish = opts.shouldDisablePublish;
```
For all these new callbacks, is it best we ensure they're optional to minimise the required number of arguments?

We could also group them into an optional `hooks` or 'lifecycle' object for readability?

matthewwalsh0 commented on app/scripts/metamask-controller.js:
```diff
@@ -4995,6 +4996,27 @@ export default class MetamaskController extends EventEmitter {
+   * @param signedEthTx - Signed ethereum transaction.
+   */
+  // eslint-disable-next-line no-unused-vars
+  _shouldDisablePublish(txMeta, signedEthTx) {
```
Rather than putting this logic and other MMI callbacks in here and bloating `MetaMaskController`, is it cleaner to put them in a separate file like `app/scripts/lib/mmi-hooks.ts`?

vinistevam replied:
Moved hooks to `app/scripts/lib/mmi-hooks.ts` and added unit test.

vinistevam commented on app/scripts/controllers/transactions/index.js:
```diff
@@ -265,9 +267,7 @@ export default class TransactionController extends EventEmitter {
     this._onBootCleanUp();
 
-    ///: BEGIN:ONLY_INCLUDE_IN(build-mmi)
     this.transactionUpdateController = opts.transactionUpdateController;
```
Removed all the references for the `transactionUpdateController`.

vinistevam commented on app/scripts/controllers/transactions/index.js:
```diff
@@ -1824,9 +1817,8 @@ export default class TransactionController extends EventEmitter {
-    ///: BEGIN:ONLY_INCLUDE_IN(build-mmi)
     // MMI does not broadcast transactions, as that is the responsibility of the custodian
-    if (txMeta.custodyStatus) {
+    if (this._shouldDisablePublish(txMeta)) {
```
Renamed `_shouldDisablePublish` to `beforePublish` and left only the `txMeta.custodyStatus` condition.

vinistevam committed:
split into several callbacks

vinistevam committed:
move callbacks to a separate file

vinistevam committed:
add get additional sign arguments

matthewwalsh0 commented on app/scripts/controllers/transactions/index.js:
```diff
@@ -1474,24 +1479,12 @@ export default class TransactionController extends EventEmitter {
-      ///: BEGIN:ONLY_INCLUDE_IN(build-mmi)
-      txMeta.custodyStatus ? txMeta : undefined,
-      ///: END:ONLY_INCLUDE_IN
+      this._getAdditionalSignArguments?.(txMeta),
```
Assuming this returns an array, would we not want to use the spread operator to include them all as additional arguments?

matthewwalsh0 commented on app/scripts/controllers/transactions/index.js:
```diff
@@ -1474,24 +1479,12 @@ export default class TransactionController extends EventEmitter {
-        fromAddress,
-      );
-
+    if (this._afterSign?.(txMeta, signedEthTx)) {
```
Would it be clearer conceptually for a return value of `false` to indicate that signing shouldn't take place as `true` implies positive which means also can continue as normal?

vinistevam replied:
indeed, makes total sense. I pushed this change.

matthewwalsh0 commented on app/scripts/controllers/transactions/index.js:
```diff
@@ -1827,9 +1820,8 @@ export default class TransactionController extends EventEmitter {
-    ///: BEGIN:ONLY_INCLUDE_IN(build-mmi)
     // MMI does not broadcast transactions, as that is the responsibility of the custodian
-    if (txMeta.custodyStatus) {
+    if (this._beforePublish?.(txMeta)) {
```
Same here, would `false` be better to skip the standard flow?

matthewwalsh0 commented on app/scripts/lib/mmi-hooks.ts:
Apologies as I know I suggested this name, but to add some context to the function names, would `mmi-transaction-hooks` be more explicit?

matthewwalsh0 commented on app/scripts/lib/mmi-hooks.ts:
```diff
@@ -0,0 +1,51 @@
+ * @param args - The list of arguments to filter.
+ */
+export const getAdditionalSignArguments = <T>(...args: T[]): T[] | T => {
+  if (args.length === 1) {
```
I'm a little confused as to what this hook is doing.

The original code fence just includes an extra argument (the `txMeta`) if there is a `custodyStatus`, so should this logic not just be:

```
return [txMeta.custodyStatus ? txMeta : undefined];
```

vinistevam replied:
Apologies. Initially, I've done a generic function but inside the MMI hooks it does make sense to only include what was between the fenced code.

matthewwalsh0 commented on app/scripts/controllers/transactions/index.js:
```diff
@@ -140,6 +140,10 @@ const METRICS_STATUS_FAILED = 'failed on-chain';
  * @param {Function} opts.hasCompletedOnboarding - Returns whether or not the user has completed the onboarding flow
  * @param {object} opts.preferencesStore
+ * @param {object} hooks
+ * @param {Function} hooks.afterSign - Determines whether to execute logic after signing a transaction.
```
Maybe:

`Additional logic to execute after signing a transaction. Return false to not change the status to signed.`

We could use a similar format for the other descriptions too?

vinistevam committed:
applied suggestion from review

vinistevam committed:
Merge branch 'develop' into 997-replace-mmi-code-fances-callbacks-arguments

vinistevam committed:
fix pipeline

matthewwalsh0 commented on app/scripts/controllers/transactions/index.js:
```diff
@@ -126,6 +126,11 @@ const VALID_UNAPPROVED_TRANSACTION_TYPES = [
  * @param {object} opts.preferencesStore
+ * @param {object} hooks
+ * @param {Function} hooks.afterSign - Additional logic to execute after signing a transaction. Return false to not change the status to signed.
+ * @param {Function} hooks.beforeApproveOnInit - Additional logic to execute logic before approving the transaction when initialized. Return false to prevent the transaction to approved.
```
```suggestion
 * @param {Function} hooks.beforeApproveOnInit - Additional logic to execute before starting an approval flow for a transaction during initialization. Return false to skip the transaction.
```

matthewwalsh0 commented on app/scripts/controllers/transactions/index.js:
```diff
@@ -126,6 +126,11 @@ const VALID_UNAPPROVED_TRANSACTION_TYPES = [
+ * @param {object} hooks
+ * @param {Function} hooks.afterSign - Additional logic to execute after signing a transaction. Return false to not change the status to signed.
+ * @param {Function} hooks.beforeApproveOnInit - Additional logic to execute logic before approving the transaction when initialized. Return false to prevent the transaction to approved.
+ * @param {Function} hooks.beforePublish - Additional logic to execute before publishing the transaction. Return false to prevent to broadcast of the transaction.
```
```suggestion
 * @param {Function} hooks.beforePublish - Additional logic to execute before publishing a transaction. Return false to prevent the broadcast of the transaction.
```

matthewwalsh0 commented on app/scripts/controllers/transactions/index.js:
```diff
@@ -126,6 +126,11 @@ const VALID_UNAPPROVED_TRANSACTION_TYPES = [
+ * @param {Function} hooks.afterSign - Additional logic to execute after signing a transaction. Return false to not change the status to signed.
+ * @param {Function} hooks.beforeApproveOnInit - Additional logic to execute logic before approving the transaction when initialized. Return false to prevent the transaction to approved.
+ * @param {Function} hooks.beforePublish - Additional logic to execute before publishing the transaction. Return false to prevent to broadcast of the transaction.
+ * @param {Function} hooks.getAdditionalSignArguments - Returns additional arguments for sign the transaction.
```
```suggestion
 * @param {Function} hooks.getAdditionalSignArguments - Returns additional arguments required to sign a transaction.
```

matthewwalsh0 commented on app/scripts/controllers/transactions/index.js:
```diff
@@ -155,6 +160,14 @@ export default class TransactionController extends EventEmitter {
     this._hasCompletedOnboarding = opts.hasCompletedOnboarding;
 
+    const { hooks } = opts ?? {};
+    this._afterSign = (...args) => hooks?.afterSign?.(...args) ?? true;
```
Very minor, but maybe slightly cleaner to not wrap the ternaries in functions. Such as:

```
this._afterSign = hooks?.afterSign ?? () => true;
```

vinistevam replied:
It was complaining about the usage of the nullish coalescing operator so I removed the wrapped function, and end up with:  `hooks?.afterSign ? hooks.afterSign : () => true`.

matthewwalsh0 commented on app/scripts/lib/mmi-hooks.ts:
```diff
@@ -0,0 +1,66 @@
+  ) => Promise<void>,
+): boolean {
+  // MMI does not broadcast transactions, as that is the responsibility of the custodian
+  if (txMeta?.custodyStatus) {
```
Very minor, but in cases like this were the bulk of the logic is nested, should we instead use an early return?

```
if (!txMeta?.custodyStatus) {
    return true;
}

txMeta.custodyId = signedEthTx.custodian_transactionId;
txMeta.custodyStatus = signedEthTx.transactionStatus;

addTransactionToWatchList(txMeta.custodyId, txMeta.txParams.from);

return false;
```

matthewwalsh0 commented on app/scripts/lib/mmi-hooks.ts:
```diff
@@ -0,0 +1,66 @@
+ */
+export function beforeTransactionPublish(txMeta: TransactionMeta): boolean {
+  // MMI does not broadcast transactions, as that is the responsibility of the custodian
+  if (txMeta?.custodyStatus) {
```
Very minor, but should we simplify to the following?

```
return !txMeta?.custodyStatus;
```

Same for `beforeTransactionApproveOnInit`?

vinistevam replied:
Perfect, done on both functions.

vinistevam committed:
Update app/scripts/controllers/transactions/index.js

Co-authored-by: Matthew Walsh <matthew.walsh@consensys.net>

vinistevam committed:
Update app/scripts/controllers/transactions/index.js

Co-authored-by: Matthew Walsh <matthew.walsh@consensys.net>

vinistevam committed:
Update app/scripts/controllers/transactions/index.js

Co-authored-by: Matthew Walsh <matthew.walsh@consensys.net>

vinistevam committed:
use early return mmi hooks

vinistevam committed:
Merge branch 'develop' into 997-replace-mmi-code-fances-callbacks-arguments

vinistevam committed:
Merge branch 'develop' into 997-replace-mmi-code-fances-callbacks-arguments

vinistevam committed:
Merge branch 'develop' into 997-replace-mmi-code-fances-callbacks-arguments

vinistevam committed:
remove underscore from alignment

vinistevam committed:
Merge branch 'develop' into 997-replace-mmi-code-fances-callbacks-arguments

vinistevam committed:
Merge branch 'develop' into 997-replace-mmi-code-fances-callbacks-arguments

vinistevam committed:
Merge branch 'develop' into 997-replace-mmi-code-fances-callbacks-arguments

shane-t committed:
Rewrite logic which checks creates noTxHashError in MMI code fence (#21562)

## **Description**

Non-custodial transactions are broken in MMI because the logic used to
manage `noTxHashError` is faulty. Previously, it would always set
noTxHash to true if the transaction was non-custodial.

## **Related issues**

Fixes: Jira ticket MMI-4063

## **Manual testing steps**

1. Use an MMI build type
2. Send a transaction with a HD or hardware wallet.


### **Before**


![image](https://github.com/MetaMask/metamask-extension/assets/1504499/2b8b40db-63f4-42ad-89b2-93faf039b3c1)


### **After**

<img width="369" alt="image"
src="https://github.com/MetaMask/metamask-extension/assets/1504499/ac88dbf5-0f33-4327-acee-53c23db3ab80">

## **Pre-merge author checklist**

- [X] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [X] I've clearly explained what problem this PR is solving and how it
is solved.
- [X] I've linked related issues
- [X] I've included manual testing steps
- [X] I've included screenshots/recordings if applicable
- [X] I’ve included tests if applicable
- [X] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [X] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
- [X] I’ve properly set the pull request status:
  - [X] In case it's not yet "ready for review", I've set it to "draft".
- [X] In case it's "ready for review", I've changed it from "draft" to
"non-draft".

## **Pre-merge reviewer checklist**

- [X] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [X] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

FrederikBolding committed:
Fix `hasApprovalFlows` condition (#21538)

## **Description**

Fix `hasApprovalFlows` condition which previously was checking whether a
function was an array, this would never be true.

sahar-fehri committed:
fix: Show number of tokens to be imported (#21521)

## **Description**
This PR fixes this github issue:
https://github.com/MetaMask/metamask-mobile/issues/7098

_1. It makes the import button show the number of tokens to be imported
rather than showing "Import All"
_2. What is the improvement/solution?_ It makes the experience look
similar to mobile

## **Manual testing steps**

_1. Step1: Add an account that has tokens with multiple tokens detected
_2. Step2: Click on Import tokens link
_3. Stem3: Notice button says Import (x)

## **Screenshots/Recordings**

_If applicable, add screenshots and/or recordings to visualize the
before and after of your change._

### **Before**


![image](https://github.com/MetaMask/metamask-extension/assets/10994169/5eb1cec4-9286-4e15-96a7-c3a1672d1a99)

### **After**


![image](https://github.com/MetaMask/metamask-extension/assets/10994169/e1967cec-fa15-4226-a62c-9921775625e6)

## **Related issues**

[_Fixes
#MMASSETS-49](https://consensyssoftware.atlassian.net/jira/software/projects/MMASSETS/boards/640/backlog?selectedIssue=MMASSETS-49)

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained:
  - [ ] What problem this PR is solving.
  - [ ] How this problem was solved.
  - [ ] How reviewers can test my changes.
- [x] I’ve indicated what issue this PR is linked to: Fixes #???
- [ ] I’ve included tests if applicable.
- [ ] I’ve documented any added code.
- [x] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
- [ ] I’ve properly set the pull request status:
  - [ ] In case it's not yet "ready for review", I've set it to "draft".
- [ ] In case it's "ready for review", I've changed it from "draft" to
"non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

vinistevam committed:
replace mmi fence code pending transaction tracker

vinistevam committed:
Merge branch 'develop' into 997-replace-mmi-code-fances-callbacks-arguments

vinistevam committed:
Merge branch 'develop' into 997-replace-mmi-code-fances-callbacks-arguments

vinistevam committed:
Merge branch 'develop' into 997-replace-mmi-code-fances-callbacks-arguments

vinistevam committed:
Merge branch 'develop' into 997-replace-mmi-code-fances-callbacks-arguments

vinistevam committed:
Merge branch 'develop' into 997-replace-mmi-code-fances-callbacks-arguments

vinistevam committed:
Merge branch 'develop' into 997-replace-mmi-code-fances-callbacks-arguments

vinistevam committed:
Merge branch 'develop' into 997-replace-mmi-code-fances-callbacks-arguments
